mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,516 @@
+                'probabilistic' but was {method_replace}"
+        )
+
+    return rz_non_cliff_selected
+
+
+def count_non_cliffords(circuit: Circuit,) -> float:
+    """Function to check how many non-Clifford gates are in a give circuit.
+
+    Args:
+        circuit: cirq.Circuit object already decomposed into the basis
+                 {Rz, Rx(pi/2), CNOT, X}
+
+    Returns:
+        number of non-Clifford gates in the given circuit.
+    """
+    operations = np.array(list(circuit.all_operations()))
+    gates = _get_gates(operations)
+    mask = np.array(
+        [isinstance(i, cirq.ops.common_gates.ZPowGate) for i in gates]
+    )
+    r_z_gates = operations[mask]
+    angles = _get_arguments(r_z_gates)
+    mask_non_cliff = ~_is_clifford_angle(angles)
+    rz_non_cliff = angles[mask_non_cliff]
+    return len(rz_non_cliff)
+
+
+def _is_clifford_angle(ang: float, tol: float = 10 ** -5,) -> bool:
+    """Function to check if a given angle is Clifford.
+    Args:
+        ang: rotation angle in the Rz gate.
+    Returns:
+        bool: True / False for Clifford or not.
+    """
+    ang = ang % (2 * np.pi)
+    closest_clifford_angle = _closest_clifford(ang)
+    if abs(closest_clifford_angle - ang) < tol:
+        return True
+    else:
+        return False
+
+
+# Vectorize function so it can take arrays of angles as its input.
+_is_clifford_angle = np.vectorize(_is_clifford_angle)
+
+
+def _closest_clifford(ang: float) -> float:
+    """Function to take angle and return the nearest Clifford angle note the
+       usage of this function is vectorized so it takes and returns arrays.
+
+    Args:
+        ang: angle in Rz gate.
+
+    Returns:
+        Clifford angle: closest clifford angle.
+    """
+    ang = ang % (2 * np.pi)
+    ang_scaled = ang / (np.pi / 2)
+    # if just one min value, return the corresponding nearest cliff.
+    if (
+        abs((ang_scaled / 0.5) - 1) > 10 ** (-6)
+        and abs((ang_scaled / 0.5) - 3) > 10 ** (-6)
+        and (abs((ang_scaled / 0.5) - 5) > 10 ** (-6))
+    ):
+        index = int(np.round(ang_scaled)) % 4
+        return CLIFFORD_ANGLES[index]
+    # if two min values (ie two cliff gates equidistant) randomly choose the
+    # cliff gate to return.
+    else:
+        index_list = [ang_scaled - 0.5, ang_scaled + 0.5]
+        index = int(choice(index_list))
+        return CLIFFORD_ANGLES[index]
+
+
+# vectorize so function can take array of angles.
+_closest_clifford = np.vectorize(_closest_clifford)
+
+
+def _random_clifford(ang: float, random_state: np.random.RandomState) -> float:
+    """Function to take angle and return the random Clifford angle note the
+       usage of this function is vectorized so it takes and returns arrays.
+
+    Args:
+        ang: angle in Rz gate.
+
+    Returns:
+        Clifford angle: closest clifford angle.
+    """
+    random_index = random_state.randint(0, 3)
+    clifford_angle = CLIFFORD_ANGLES[random_index]
+    return clifford_angle
```

Comment:
`ang` is not used? What is this function supposed to do?

Reply:
this function returns random angles as long as the array of angles input, hence the argument ang. But its functionality is a little hidden. Thats true.